File: Leetcode Daily Challenge/013-July-2024/005-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/




# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def nodesBetweenCriticalPoints(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: List[int]
        """
        return self.optimal(head)

    # TC : O(N)
    # SC : O(1)
    def optimal(self, head):
            first_critical_pt = 0
            prev_critical_pt = 0
            count_of_nodes = 2
            minima = float('inf')

            while head.next.next:
                if ((head.next.val < head.val and head.next.val < head.next.next.val) or
                    (head.next.val > head.val and head.next.val > head.next.next.val)):

                    if first_critical_pt == 0:
                        first_critical_pt = count_of_nodes
                        prev_critical_pt = count_of_nodes
                    else:
                        minima = min(minima, count_of_nodes - prev_critical_pt)
                        prev_critical_pt = count_of_nodes

                head = head.next
                count_of_nodes += 1

            if minima == float('inf'):
                return [-1, -1]

            maxima = prev_critical_pt - first_critical_pt
            return [minima, maxima]

File: Leetcode Daily Challenge/013-July-2024/005-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
// https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/



import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        return optimal(head);
    }

    // TC : O(N) Where N = No. of Nodes
    // SC : O(1)
    private static int[] optimal(ListNode head) {    
        int firstCriticalPt = 0;
        int prevCriticalPt = 0;
        int countOfNodes = 2;
        int minima = Integer.MAX_VALUE;

        while (head.next.next != null) {
            if ((head.next.val < head.val && head.next.val < head.next.next.val) ||
                (head.next.val > head.val && head.next.val > head.next.next.val)) {

                if (firstCriticalPt == 0) {
                    firstCriticalPt = countOfNodes;
                    prevCriticalPt = countOfNodes;
                } else {
                    minima = Math.min(minima, countOfNodes - prevCriticalPt);
                    prevCriticalPt = countOfNodes;
                }
            }
            head = head.next;
            countOfNodes++;
        }

        if (minima == Integer.MAX_VALUE)
            return new int[]{-1, -1};

        int maxima = prevCriticalPt - firstCriticalPt;
        return new int[]{minima, maxima};
    }
}